src/core/lib/security/transport/security_connector.c

```diff
@@ -511,14 +537,107 @@ static void ssl_channel_add_handshakers(grpc_exec_ctx *exec_ctx,
+    gpr_log(GPR_ERROR,
+            "Server credentials callback returned invalid (NULL) credentials.");
+    return;
+  }
+
+  if (0 != strcmp(GRPC_CHANNEL_CREDENTIALS_TYPE_SSL, server_creds->type)) {
+    gpr_log(GPR_ERROR,
+            "Server credentials callback returned invalid, non-SSL "
+            "credentials type (%s).",
+            server_creds->type);
+    return;
+  }
+
+  gpr_log(GPR_DEBUG, "Using new server credentials with server_creds=%p.",
+          server_creds);
+
+  ssl_server_creds = (grpc_ssl_server_credentials *)server_creds;
+
+  alpn_protocol_strings = fill_alpn_protocol_strings();
+  result = tsi_create_ssl_server_handshaker_factory_ex(
+      ssl_server_creds->config.pem_key_cert_pairs,
+      ssl_server_creds->config.num_key_cert_pairs,
+      ssl_server_creds->config.pem_root_certs,
+      get_tsi_client_certificate_request_type(
+          ssl_server_creds->config.client_certificate_request),
+      ssl_cipher_suites(), alpn_protocol_strings, (uint16_t)num_alpn_protocols,
+      &new_handshaker_factory);
+  gpr_free((void *)alpn_protocol_strings);
+
+  if (result == TSI_OK) {
+    tsi_ssl_server_handshaker_factory_unref(sc->server_handshaker_factory);
+    sc->server_handshaker_factory = new_handshaker_factory;
+  } else {
+    gpr_log(GPR_ERROR, "Handshaker factory creation failed with %s.",
+            tsi_result_to_string(result));
+  }
+}
+
+/* Attempts to reload the server credentials if a callback is available. Current
+ * credentials will continue to be used if the callback returns an error. */
+static void reload_ssl_server_credentials(
+    grpc_ssl_server_security_connector *sc) {
+  grpc_server_credentials *server_creds = NULL;
+  grpc_get_server_credentials_result cb_result;
+
+  if (sc == NULL || sc->base.get_server_credentials_cb == NULL) return;
+
+  cb_result = sc->base.get_server_credentials_cb(
+      &server_creds, sc->base.get_server_credentials_cb_arg);
+
+  if (cb_result == GRPC_GET_SERVER_CREDENTIALS_UNCHANGED) {
+    gpr_log(GPR_DEBUG, "No change in SSL server credentials.");
+  } else if (cb_result == GRPC_GET_SERVER_CREDENTIALS_NEW) {
+    replace_server_handshaker_factory(sc, server_creds);
+  } else {
+    // Log error, continue using previously-loaded credentials.
+    gpr_log(GPR_ERROR, "Failed loading new server credentials, error code %d.",
+            cb_result);
+  }
+
+  if (server_creds) {
+    grpc_server_credentials_release(server_creds);
+  }
+}
+
 static void ssl_server_add_handshakers(grpc_exec_ctx *exec_ctx,
                                        grpc_server_security_connector *sc,
                                        grpc_handshake_manager *handshake_mgr) {
   grpc_ssl_server_security_connector *c =
       (grpc_ssl_server_security_connector *)sc;
   // Instantiate TSI handshaker.
   tsi_handshaker *tsi_hs = NULL;
-  tsi_result result = tsi_ssl_server_handshaker_factory_create_handshaker(
+  tsi_result result;
+
+  reload_ssl_server_credentials(c);
```

Comment:
naming nit: perhaps `maybe_reload_ssl_server_credentials()` to indicate that the reload is a "maybe"?

Reply:
Sure. Will update.